### PR TITLE
[JENA-2312] Use the service endpoint in the URL used by the UI to que…

### PR DESCRIPTION
**Pull request by kinow**
…ry the backend

### CODE REVIEW

File: jena-fuseki2/jena-fuseki-webapp/pom.xml [hunk 1]
```diff
@@ -57,6 +57,12 @@
       <version>4.5.0-SNAPSHOT</version>
     </dependency>
     
+    <dependency>
+      <groupId>org.apache.jena</groupId>
+      <artifactId>jena-geosparql</artifactId>
+      <version>4.5.0-SNAPSHOT</version>
+    </dependency>
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-webapp/pom.xml:64]: Ops, I will remove this change. Only added it to test the GeoSparql pull request :+1:

**Comment by kinow** [comment]
I think the `/` endpoint, used as a service for `query` cannot handle uploads? I'm extracting the upload URL from the `gsp-rw` service. Does that sound correct?

And will display a message to the user about the missing service. What others think?

image

**Comment by kinow** [comment]
Sort of fixed. Pending:

- [ ] test other views and components
- [ ] test both cases, where we have multiple services configured for the dataset, and also a dataset with just the query service as in the geosparql example
- [ ] confirm that the `gsp-rw` is necessary to upload data
- [ ] confirm what is necessary for the edit graph
- [ ] check if unit tests are working after this change

Now :sleeping_bed:

**Comment by kinow** [comment]
@afs , are you able to answer these two?

>confirm that the gsp-rw is necessary to upload data

The GeoSPARQL example from the documentation doesn't seem to allow uploading data? Or is it a bug in the UI?

```ttl
<#service> rdf:type fuseki:Service;
    fuseki:name "geo";
    fuseki:endpoint [ fuseki:operation fuseki:query; ] ;
    fuseki:dataset <#geo_ds> .

<#geo_ds> rdf:type geosparql:geosparqlDataset ;
    geosparql:spatialIndexFile     "DB/spatial.index";
    geosparql:dataset <#baseDataset> ;
    .

<#baseDataset> rdf:type tdb2:DatasetTDB2 ;
    tdb2:location "DB/" ;
    .
```

>confirm what is necessary for the edit graph

Similar to the previous question, is there a service that must be enabled so that the UI can edit the graph in the UI?

Thanks!

**Comment by kinow** [comment]
Still not able to edit the data for the GeoSPARQL dataset.

image

I guess we also need the `gsp-rw` service enabled for that? If so I will disable the UI when the selected dataset is missing that option.

The rest of the UI is looking OK, so I think the change here will fix the issue of the service endpoint. Ready for an initial review, pending some questions/checks.

**Comment by afs** [comment]
> confirm that the gsp-rw is necessary to upload data

Yes.

The "gsp-rw" includes the extension for loading quads.

There is another server "upload", which is no longer added in the default setup. "gsp-rw" calls the "upload" functionality when it receives a file without the GSP `?default` and `?graph=` parameters so it is a superset.

**Comment by kinow** [comment]
Perfect, I'll update the Edit view to tell the user when it's not usable due to the missing service. Thanks @afs!

**Review by afs — APPROVED** [review]

File: jena-fuseki2/jena-fuseki-webapp/pom.xml (hunk 1 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-webapp/pom.xml:64]: Shall I merge then remove this? I'm clearing up prior to 4.5 so in the area.

**Comment by kinow** [comment]
Using the GeoSPARQL docs dataset, and a normal TDB2 dataset to test. This PR changes `Upload` and `Edit` only. Here's a preview of these views for the two aforementioned datasets.

image

image

image

image

The views do not display the forms when there are no `gsp-rw` graphs configured. @afs all done now, if that's looking OK it should be ready to be merged. Thanks!

File: jena-fuseki2/jena-fuseki-webapp/pom.xml (hunk 1 above)
Comment (kinow) [reply jena-fuseki2/jena-fuseki-webapp/pom.xml]: Reverted the change :+1: But thanks for offering to handle this @afs :bow:

**Review by rvesse — APPROVED** [review]

File: jena-fuseki2/jena-fuseki-ui/src/views/dataset/Edit.vue [hunk 2]
```diff
@@ -24,16 +24,20 @@
           <b-card-header header-tag="nav">
             <Menu :dataset-name="datasetName" />
           </b-card-header>
-          <b-card-body>
+          <b-card-body v-if="!this.services['gsp-rw'] || this.services['gsp-rw'].length === 0">
+            <b-alert show variant="warning">No service for Graph Store Protocol configured</b-alert>
```
Comment (afs) [inline jena-fuseki2/jena-fuseki-ui/src/views/dataset/Edit.vue:28]: Uses may not associate GSP with thier task of adding data. Maybe working like:

```suggestion
            <b-alert show variant="warning">No service for adding data available. The Graph Store Protocol service should be configured to allow adding data.</b-alert>
```

**Review by afs — APPROVED** [review]

File: jena-fuseki2/jena-fuseki-ui/src/views/dataset/Edit.vue (hunk 2 above)
Comment (kinow) [reply jena-fuseki2/jena-fuseki-ui/src/views/dataset/Edit.vue:28]: Will copy the text in Upload.vue too. Thanks!

**Pull request merged**